Changed tesla state_class from TOTAL_INCREASING to handle slight decreases

ptarjan opened the pull request:
<img width="3154" height="363" alt="image" src="https://github.com/user-attachments/assets/1d7e67c7-a32c-46e3-b963-f6a3c0e8faf3" />

The root cause was that the Tesla API can report small decreases in the charge_energy_added value due to recalculations or precision issues (3.82 → 3.76 in your case). The TOTAL_INCREASING state class expects values to only increase or reset to zero, which doesn't match the actual API behavior. 

I also update the message to make it more clear: Changed "Triggered by state 3.76 (3.82)" to "Triggered by state 3.76 (previous state: 3.82)"


<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [ ] I understand the code I am submitting and can explain how it works.
- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.
- [ ] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Bre77 commented:
Doesn't this make it impossible to aggregate the total energy over multiple days?

epenet reviewed (commented):
I think the new state class is wrong - if it is incorrect it should be completely removed.
If it's a rounding issue it should be handled/adjusted in the integration to ignore these discrepencies

ptarjan commented:
It seems mostly right. It goes up most of the times, just sometimes it goes down a tad. I assume it is an out-of-order distributed system message problem, or maybe a rounding error in measurement (less likely). How should the integration deal with those?

ptarjan committed:
Handle tesla_fleet charge_energy_added dips in integration

Use TOTAL state class with last_reset handling instead of TOTAL_INCREASING.
This properly handles the Tesla API's occasional value recalculations:

- Changed state_class from TOTAL_INCREASING to TOTAL
- Added last_reset attribute that updates when a new charging session starts
  (detected when value drops significantly, < 50% of previous)
- Accept all values from Tesla API as correct (no filtering)

This approach allows the statistics system to properly track energy across
charging sessions while accepting Tesla's recalculated values as accurate.

joostlek commented:
So I can't see what was pushed before, but looking at the code now, if the service does some recalculations and adjust to a bit lower number, it's that exactly what total is?

Bre77 commented:
> Doesn't total also reset when it hits 0?

From my read of the documentation, no, you have to set last_reset too. This PR would need to account for this.

```For sensors with state class total, the last_reset attribute can optionally be set to gain manual control of meter cycles. The sensor's state when it's first added to Home Assistant is used as an initial zero-point. When a last_reset changes, the zero-point will be set to 0. If last_reset is not set, the sensor's value when it was first added is used as the zero-point when calculating sum statistics.```

ptarjan commented on homeassistant/components/sensor/recorder.py:
```diff
@@ -416,8 +416,8 @@ def warn_dip(
                 "Entity %s %shas state class total_increasing, but its state is not"
-                " strictly increasing. Triggered by state %s (%s) with last_updated set"
-                " to %s. Please %s"
+                " strictly increasing. Triggered by state %s (previous state: %s) with"
```
Addressed. The sensor/recorder.py changes have been removed from this PR. It now only touches tesla_fleet files.

ptarjan commented:
@epenet Thanks for the review. To clarify the current approach:

- The state class was changed to `TOTAL` per @Bre77's recommendation (https://github.com/home-assistant/core/pull/158749#issuecomment-3830528674), since the Tesla API can report small decreases in `charge_energy_added` due to recalculations (e.g. 3.82 -> 3.76).
- `last_reset` is now set when a new charging session is detected (identified by a >50% drop in value), which resets the zero-point for sum statistics as the `TOTAL` docs describe.
- For small rounding decreases within a session, `TOTAL` will accept them without error -- unlike `TOTAL_INCREASING`, which would reset statistics on any decrease. The small negative deltas in the recorder represent corrections from Tesla's API rather than actual energy decreases, so they should be acceptable.

That said, I understand your concern that the state class might be papering over the issue. Would you prefer the integration to also clamp values within a session to prevent any decrease (e.g. `max(new_value, current_value)`)? Happy to add that if it's the preferred approach.

Bre77 commented:
Hey @ptarjan sorry I didnt let you know sooner, I have addressed this issue across `tessie`, `teslemetry`, and `tesla_fleet` and across a number of entities in https://github.com/home-assistant/core/pull/162528. Ill do a quick check to see if there is any different in our implementations.
